File: python/course-repo/solution/03-dedup-via-id-reuse/src/webhooks/send_double.py
"""Fire start_activity TWICE with the same id, back-to-back.

The second call uses ActivityIDConflictPolicy.USE_EXISTING so it
quietly returns the existing handle instead of erroring.
"""

import asyncio
import sys
from datetime import timedelta

from temporalio.client import Client
from temporalio.common import ActivityIDConflictPolicy

from .activities import deliver_webhook
from .shared import WEBHOOK_RECEIVER_URL, TASK_QUEUE, WebhookDelivery


async def start(client: Client, event_id: str, label: str):
    """Try to start one activity with the given event_id."""
    print(f"[{label}] start_activity id=deliver-{event_id}")
    try:
        handle = await client.start_activity(
            deliver_webhook,
            args=[WebhookDelivery(
                url=WEBHOOK_RECEIVER_URL,
                payload={"event_id": event_id, "type": "dup_test"},
                event_id=event_id,
            )],
            id=f"deliver-{event_id}",
            task_queue=TASK_QUEUE,
            start_to_close_timeout=timedelta(seconds=30),
            # USE_EXISTING: second submission with the same id returns the
            # existing handle instead of erroring. Server-side dedup; the
            # duplicate never reaches a Worker.
            id_conflict_policy=ActivityIDConflictPolicy.USE_EXISTING,
        )
        print(f"[{label}] handle ok (run_id={handle.run_id})")
        return handle
    except Exception as e:
        print(f"[{label}] FAILED: {type(e).__name__}: {e}")
        return None


async def main(event_id: str) -> None:
    client = await Client.connect("localhost:7233")
    h1 = await start(client, event_id, "call-1")
    h2 = await start(client, event_id, "call-2")

    if h1 is not None:
        await h1.result()
        print("[call-1] activity completed")
    if h2 is not None:
        await h2.result()
        print("[call-2] activity completed")


if __name__ == "__main__":
    event_id = sys.argv[1] if len(sys.argv) > 1 else "evt_dup"
    asyncio.run(main(event_id))

File: java/course-repo/solution/03-dedup-via-id-reuse/src/main/java/webhooks/SendDouble.java
package webhooks;

import io.temporal.api.enums.v1.ActivityIdConflictPolicy;
import io.temporal.client.ActivityClient;
import io.temporal.client.ActivityClientOptions;
import io.temporal.client.ActivityHandle;
import io.temporal.client.StartActivityOptions;
import io.temporal.serviceclient.WorkflowServiceStubs;

import java.time.Duration;
import java.util.HashMap;
import java.util.Map;

/**
 * Calls start() TWICE with the same Activity ID, back-to-back.
 *
 * The conflict policy is set to USE_EXISTING, so the second call quietly
 * returns a handle to the existing run instead of erroring. Server-side dedup
 * before any Worker sees the duplicate job.
 */
public class SendDouble {

  static ActivityHandle<Integer> start(
      ActivityClient client, String eventId, String label) {
    System.out.println("[" + label + "] start activityId=deliver-" + eventId);

    Map<String, Object> payload = new HashMap<>();
    payload.put("eventId", eventId);
    payload.put("type", "dup_test");
    WebhookDelivery req = new WebhookDelivery(Shared.WEBHOOK_RECEIVER_URL, payload, eventId);

    StartActivityOptions options =
        StartActivityOptions.newBuilder()
            .setId("deliver-" + eventId)
            .setTaskQueue(Shared.TASK_QUEUE)
            .setStartToCloseTimeout(Duration.ofSeconds(30))
            // USE_EXISTING: a second start with the same Activity ID returns the
            // existing handle instead of erroring. Server-side dedup before any
            // Worker sees the job.
            .setIdConflictPolicy(
                ActivityIdConflictPolicy.ACTIVITY_ID_CONFLICT_POLICY_USE_EXISTING)
            .build();

    try {
      ActivityHandle<Integer> handle =
          client.start(WebhookActivities.class, WebhookActivities::deliverWebhook, options, req);
      System.out.println("[" + label + "] handle ok (activityId=deliver-" + eventId + ")");
      return handle;
    } catch (Exception err) {
      System.out.println("[" + label + "] FAILED: " + err);
      return null;
    }
  }

  public static void main(String[] args) {
    String eventId = args.length > 0 ? args[0] : "evt_dup";

    WorkflowServiceStubs service = WorkflowServiceStubs.newLocalServiceStubs();
    ActivityClient client =
        ActivityClient.newInstance(
            service, ActivityClientOptions.newBuilder().setNamespace("default").build());

    ActivityHandle<Integer> h1 = start(client, eventId, "call-1");
    // The second call simulates the upstream retry/replay arriving right after the first.
    ActivityHandle<Integer> h2 = start(client, eventId, "call-2");

    if (h1 != null) {
      h1.getResult();
      System.out.println("[call-1] activity completed");
    }
    if (h2 != null) {
      h2.getResult();
      System.out.println("[call-2] activity completed");
    }

    System.exit(0);
  }
}
